### ISSUE TITLE
Scaling "Docker For AWS" nodes

### ISSUE THREAD

**Issue by vfarcic** [thomasjpfan/docker-scaler#3]
I'm using the following snippet (Jenkins Pipeline Groovy) to scale "Docker For AWS" nodes.

```groovy
        git "https://github.com/vfarcic/docker-aws-cli.git"
        script {
          def asgName = sh(
            script: "source /run/secrets/aws && docker-compose run --rm asg-name",
            returnStdout: true
          ).trim()
          if (asgName == "") {
            error "Could not find auto-scaling group"
          }
          def asgDesiredCapacity = sh(
            script: "source /run/secrets/aws && ASG_NAME=${asgName} docker-compose run --rm asg-desired-capacity",
            returnStdout: true
          ).trim().toInteger()
          def asgNewCapacity = asgDesiredCapacity + scale.toInteger()
          sh "source /run/secrets/aws && ASG_NAME=${asgName} ASG_DESIRED_CAPACITY=${asgNewCapacity} docker-compose run --rm asg-update-desired-capacity"
          echo "Changed the number of workers from ${asgDesiredCapacity} to ${asgNewCapacity}"
        }
```

It does the following.

1) Get the name of the auto-scaling group associated with worker nodes.
2) Use 1) to get the current number of auto-scaling group desired capacity.
3) Calculate the new number of desired capacity (e.g. `current + 1` or `current - 2`).
4) Update the auto-scaling group to the new desired capacity.

Do you think that a similar logic would be a good addition to the project?

**Comment by thomasjpfan** [comment]
Expanding the project to scale physical nodes looks good. There is even a Go SDK to use: http://docs.aws.amazon.com/sdk-for-go/api/service/autoscaling/.

**Comment by thomasjpfan** [comment]
I just added AWS scaling that follows your Jenkins pipeline. For now I'm using an env variable, `AWS_LAUNCH_CONFIGURATION_NAME`, as the `LaunchConfigurationName`. The AWS scaling code is [here](https://github.com/thomasjpfan/docker-scaler/blob/master/service/awsscaler.go#L46-L76). The POST request to start a node scaling event is:
```bash
$ curl -X POST localhost:8080/scale?nodesOn=aws&delta=1
```
A docker swarm can be connected between multiple providers. With the `nodesOn` parameter, the caller can decide where to expand their worker nodes. 

Like your Jenkins script, the node scaler will source an env file at `/run/secrets/aws` by default. The env file should have the form:
```bash
export AWS_ACCESS_KEY_ID=[xxxxx]
export AWS_SECRET_ACCESS_KEY=[xxxxxx]
export AWS_DEFAULT_REGION=[xxxxxx]
```
I have not tested the AWS scaling feature on an live AWS cluster yet. There may or may not be bugs!

**Comment by vfarcic** [comment]
Great!

I'll be traveling for the rest of this week. I'll go through it on Monday.

**Comment by thomasjpfan** [comment]
I just finished testing on an live AWS cluster, and verified that the latest version is able to scale EC2 nodes. A sample Docker stack file can be found [here](https://github.com/thomasjpfan/docker-scaler/blob/master/scripts/docker-compose-aws.yml).

I have made some changes to the `POST` request. For example, to scale up worker nodes:
```bash
$ curl -X POST http://$CLUSTER_DNS:8080/scale?nodesOn=aws&delta=1&type=worker
```
Or to scale down manager nodes:
```bash
$ curl -X POST http://$CLUSTER_DNS:8080/scale?nodesOn=aws&delta=-1&type=manager
```
The `/run/secrets/aws` env file is of the form:
```bash
export AWS_ACCESS_KEY_ID=[xxxxx]
export AWS_SECRET_ACCESS_KEY=[xxxxxx]
export AWS_REGION=[xxxxxx]
```
where `AWS_REGION` is optional. The AWS go client will read from `AWS_DEFAULT_REGION` if `AWS_REGION` is not given.

Instead of using the configuration name, I used the manager group name and the worker group name, when interacting with the auto-scaling groups.
